File: tests/repos/typescript-calculator/src/operations/basic.ts
/**
 * Basic arithmetic operations module.
 */

// Named imports
import {
    NumberInput,
    OperatorSymbol,
    OperationResult,
    CalculationError,
    ErrorCode,
    Ok,
    Err,
    type Result,
} from './types.js';

// Default import
import Decimal from 'decimal.js';

// Re-export
export { CalculationError, ErrorCode } from './types.js';

/**
 * Add two numbers.
 */
export function add(a: number, b: number): number {
    return a + b;
}

/**
 * Subtract b from a.
 */
export function subtract(a: number, b: number): number {
    return a - b;
}

/**
 * Multiply two numbers.
 */
export function multiply(a: number, b: number): number {
    return a * b;
}

/**
 * Divide a by b.
 * Returns Result type for safe error handling.
 */
export function divide(a: number, b: number): Result<number, CalculationError> {
    if (b === 0) {
        return new Err(
            new CalculationError(
                'Division by zero',
                ErrorCode.DivisionByZero,
                'divide',
                [a, b]
            )
        );
    }
    return new Ok(a / b);
}

/**
 * Calculate power.
 */
export function power(base: number, exponent: number): number {
    return Math.pow(base, exponent);
}

/**
 * Calculate modulo.
 */
export function modulo(a: number, b: number): Result<number, CalculationError> {
    if (b === 0) {
        return new Err(
            new CalculationError(
                'Modulo by zero',
                ErrorCode.DivisionByZero,
                'modulo',
                [a, b]
            )
        );
    }
    return new Ok(a % b);
}

/**
 * Sum all provided numbers using rest parameters.
 */
export function sum(...numbers: number[]): number {
    return numbers.reduce((acc, n) => acc + n, 0);
}

/**
 * Calculate product of all numbers.
 */
export function product(...numbers: number[]): number {
    if (numbers.length === 0) return 0;
    return numbers.reduce((acc, n) => acc * n, 1);
}

/**
 * Get operation function by operator symbol.
 * Demonstrates function as return type.
 */
export function getOperationBySymbol(
    operator: OperatorSymbol
): ((a: number, b: number) => number) | null {
    const operations: Record<OperatorSymbol, (a: number, b: number) => number> = {
        '+': add,
        '-': subtract,
        '*': multiply,
        '/': (a, b) => a / b,
        '**': power,
        '%': (a, b) => a % b,
    };

    return operations[operator] ?? null;
}

/**
 * Create a curried operation function.
 */
export function createOperation(
    operator: OperatorSymbol
): (a: number) => (b: number) => number {
    const op = getOperationBySymbol(operator);
    if (!op) {
        throw new CalculationError(
            `Unknown operator: ${operator}`,
            ErrorCode.InvalidOperation
        );
    }
    return (a: number) => (b: number) => op(a, b);
}

/**
 * Precise decimal arithmetic using Decimal.js library.
 */
export function preciseAdd(a: NumberInput, b: NumberInput): string {
    return new Decimal(a.toString()).plus(new Decimal(b.toString())).toString();
}

export function preciseMultiply(a: NumberInput, b: NumberInput): string {
    return new Decimal(a.toString()).times(new Decimal(b.toString())).toString();
}

export function preciseDivide(
    a: NumberInput,
    b: NumberInput,
    precision: number = 20
): string {
    Decimal.set({ precision });
    return new Decimal(a.toString()).dividedBy(new Decimal(b.toString())).toString();
}

// Lambda/arrow function expressions
export const negate = (x: number): number => -x;
export const absolute = (x: number): number => Math.abs(x);
export const square = (x: number): number => x * x;
export const sqrt = (x: number): number | null => (x >= 0 ? Math.sqrt(x) : null);

// IIFE (Immediately Invoked Function Expression)
export const PI_SQUARED = (() => Math.PI * Math.PI)();

// Higher-order function
export function compose<T>(
    ...fns: Array<(arg: T) => T>
): (arg: T) => T {
    return (arg: T) => fns.reduceRight((acc, fn) => fn(acc), arg);
}

export function pipe<T>(
    ...fns: Array<(arg: T) => T>
): (arg: T) => T {
    return (arg: T) => fns.reduce((acc, fn) => fn(acc), arg);
}

// Generic higher-order functions
export function map<T, U>(fn: (item: T) => U): (arr: T[]) => U[] {
    return (arr: T[]) => arr.map(fn);
}

export function filter<T>(predicate: (item: T) => boolean): (arr: T[]) => T[] {
    return (arr: T[]) => arr.filter(predicate);
}

export function reduce<T, U>(
    reducer: (acc: U, item: T) => U,
    initial: U
): (arr: T[]) => U {
    return (arr: T[]) => arr.reduce(reducer, initial);
}

// Batch operation using array methods
export function batchOperation(
    pairs: [number, number][],
    operator: OperatorSymbol
): OperationResult[] {
    const op = getOperationBySymbol(operator);
    if (!op) {
        return pairs.map(([a, b]) => ({
            value: 0,
            operation: operator,
            inputs: [a, b],
            success: false,
            error: new CalculationError(
                'Unknown operator',
                ErrorCode.InvalidOperation
            ),
        }));
    }

    return pairs.map(([a, b]) => {
        try {
            const value = op(a, b);
            return {
                value,
                operation: operator,
                inputs: [a, b],
                success: true,
            };
        } catch (e) {
            return {
                value: 0,
                operation: operator,
                inputs: [a, b],
                success: false,
                error:
                    e instanceof CalculationError
                        ? e
                        : new CalculationError(
                              String(e),
                              ErrorCode.InvalidOperation
                          ),
            };
        }
    });
}

// Default export
export default {
    add,
    subtract,
    multiply,
    divide,
    power,
    modulo,
    sum,
    product,
};

File: tests/repos/python-calculator/operations/basic.py
"""Basic arithmetic operations module."""

from typing import Union, Optional
from functools import reduce
from decimal import Decimal, ROUND_HALF_UP
import math

# Type alias
Number = Union[int, float, Decimal]


def add(a: Number, b: Number) -> Number:
    """Add two numbers."""
    return a + b


def subtract(a: Number, b: Number) -> Number:
    """Subtract b from a."""
    return a - b


def multiply(a: Number, b: Number) -> Number:
    """Multiply two numbers."""
    return a * b


def divide(a: Number, b: Number) -> Optional[Number]:
    """
    Divide a by b.

    Returns None if division by zero is attempted.
    """
    if b == 0:
        return None
    return a / b


def sum_all(*args: Number) -> Number:
    """Sum any number of arguments using reduce."""
    return reduce(lambda x, y: x + y, args, 0)


def product_all(*args: Number) -> Number:
    """Multiply any number of arguments."""
    return reduce(multiply, args, 1)


def safe_divide(a: Number, b: Number, default: Number = 0) -> Number:
    """Safe division with default value on error."""
    result = divide(a, b)
    return default if result is None else result


def round_result(value: Number, precision: int = 2) -> Decimal:
    """Round a number to specified decimal places."""
    decimal_value = Decimal(str(value))
    quantize_str = '0.' + '0' * precision
    return decimal_value.quantize(Decimal(quantize_str), rounding=ROUND_HALF_UP)


def batch_operation(
    operation: str,
    numbers: list[Number]
) -> Optional[Number]:
    """
    Apply an operation to a list of numbers.

    Uses match statement (Python 3.10+) for operation dispatch.
    """
    if not numbers:
        return None

    match operation:
        case 'add' | 'sum':
            return sum_all(*numbers)
        case 'multiply' | 'product':
            return product_all(*numbers)
        case 'max':
            return max(numbers)
        case 'min':
            return min(numbers)
        case 'average' | 'avg':
            return sum_all(*numbers) / len(numbers)
        case _:
            return None


# Lambda functions for single operations
negate = lambda x: -x
absolute = lambda x: abs(x)
square = lambda x: x ** 2
sqrt = lambda x: math.sqrt(x) if x >= 0 else None


# Higher-order function
def create_operation(operator: str):
    """Factory function returning operation lambdas."""
    operations = {
        '+': lambda a, b: a + b,
        '-': lambda a, b: a - b,
        '*': lambda a, b: a * b,
        '/': lambda a, b: a / b if b != 0 else None,
        '**': lambda a, b: a ** b,
        '%': lambda a, b: a % b if b != 0 else None,
    }
    return operations.get(operator, lambda a, b: None)
